src/data/UserD2Repository.ts

```diff
@@ -63,4 +63,73 @@ export class UserD2Repository implements UserRepository {
             .compact()
             .value();
     }
+
+    async getByUserName(username: string): Async<User | undefined> {
+        const { users } = await this.api.metadata
+            .get({
+                users: {
+                    fields: {
+                        id: true,
+                        email: true,
+                        userCredentials: { username: true },
+                    },
+                    filter: {
+                        "userCredentials.username": { eq: username },
+                    },
+                },
+            })
+            .getData();
+
+        return _(users)
+            .map(d2User => ({
+                id: d2User.id,
+                email: d2User.email,
+                username: d2User.userCredentials.username,
+            }))
+            .first();
+    }
+
+    async updateUserName(oldUserName: string, newUserName: string): Async<string> {
```

Comment:
`Async<string>` is unspecific, so one must inspected to see what this string represent (an error). We have two approaches: 1) use some kind of `Either` (success/error) type, or, 2) since an Async/Promise already has an implicit error branch (by throwing an exception, equivalent to Future.error). The second way is the more simple: `: Async<void>`. And on error -> `throw new Error("Some error")`.